File: semestr-8/Jezyk C#/alexander/Csharp/Archiwum/2021s/Id06IO1/MatrixOperation/MatrixOperation/Matrix.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MatrixOperation
{
    public class Matrix
    {
        private double[,] mat;
        public Matrix(string text)
        {
            mat = Parse(text);
        }
        public Matrix(int rows,int cols)
        {
            mat = new double[rows,cols];
        }
        public int rows { get { return mat.GetLength(0); } }
        public int cols { get { return mat.GetLength(1); } }
        private static string NormSeparator(string text)
        {
            return 
                text
                    .Replace(",", System.Globalization.NumberFormatInfo.CurrentInfo.NumberDecimalSeparator)
                    .Replace(".", System.Globalization.NumberFormatInfo.CurrentInfo.NumberDecimalSeparator)
                ;
        }
        private static double[,] Parse(string text)
        {
            double[,] mat=null;
            string[] rows = text.Split("\r\n\v".ToArray(), StringSplitOptions.RemoveEmptyEntries);
            for(int r=0;r<rows.Length;++r)
            {
                string[] cols = rows[r].Split(" \t".ToArray(), StringSplitOptions.RemoveEmptyEntries);
                for(int c=0;c<cols.Length;++c)
                {
                    if ((r == 0) && (c == 0)) mat = new double[rows.Length, cols.Length];
                    else if (cols.Length != mat.GetLength(1)) throw new Exception("Rożne rozmiary wierszy");
                    mat[r, c] = double.Parse(NormSeparator(cols[c]));
                }
            }
            return mat;
        }
        public static Matrix operator +(Matrix a, Matrix b) { return add(a, b); }
        public static Matrix operator *(Matrix a, Matrix b) { return mul(a, b); }
        public static Matrix add(Matrix a,Matrix b)
        {
            int rows = a.rows,cols = a.cols;
            if (rows != b.rows) throw new Exception("Rożna ilość wierszy");
            if (cols != b.cols) throw new Exception("Rożna ilość kolumn");
            Matrix m = new Matrix(rows, cols);
            for (int r = 0; r < rows; ++r)
            {
                for (int c = 0; c < cols; ++c)
                {
                    m.mat[r, c] = a.mat[r, c] + b.mat[r, c];
                }
            }
            return m;
        }
        public double this[int r, int c]
        {
            get { return mat[r, c]; }
            set { mat[r, c] = value; }
        }
        public static Matrix mul(Matrix a, Matrix b)
        {
            int rows = a.rows, cols = b.cols, size = a.cols;
            if (size != b.rows) throw new Exception("Nieodpowiednie rozmiary macierzy");
            Matrix m = new Matrix(rows, cols);
            for (int r = 0; r < rows; ++r)
            {
                for (int c = 0; c < cols; ++c)
                {
                    double sum = 0;
                    for (int s = 0; s < size; ++s) sum += a[r, s] * b[s, c];
                    //m.mat[r, c] = sum;
                    m[r, c] = sum;
                }
            }
            return m;
        }
        public static Matrix solve(Matrix m, Matrix y)
        {
            int rows = m.rows, mcols = m.cols, ycols = y.cols, cols = mcols + ycols;
            if (rows != m.cols) throw new Exception("Pierwsza macierz musi być kwadratowa");
            if (rows != y.rows) throw new Exception("Ilość wierszy obu macierzy musi być jednakowa");
            Matrix tmp = new Matrix(rows, cols);
            for (int r = 0; r < rows; ++r)
            {
                for (int c = 0; c < mcols; ++c) tmp[r, c] = m[r,c];
                for (int c = 0; c < ycols; ++c) tmp[r, mcols + c] = y[r, c];
            }
            for (int d = 0; d < rows; ++d)
            {
                double div = tmp[d, d];
                for (int r = 0; r < rows; ++r)
                {
                    if (r == d) continue;
                    double mul = tmp[r, d];
                    for (int c = 0; c < cols; ++c) tmp[r, c] -= tmp[d, c] * mul / div;
                }
            }
            for (int d = 0; d < rows; ++d)
            {
                double div = tmp[d, d];
                for (int c = 0; c < cols; ++c) tmp[d, c] /= div;
            }
            y = new Matrix(rows, ycols);
            for (int r = 0; r < rows; ++r)
            {
                for (int c = 0; c < ycols; ++c) y[r, c] = tmp[r, mcols + c];
            }
            return y;
        }
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            for (int r = 0; r < rows; ++r)
            {
                for (int c = 0; c < cols; ++c)
                {
                    if (c > 0) sb.Append('\t');
                    sb.Append(string.Format("{0,10:F5}", mat[r, c]));
                }
                sb.Append(Environment.NewLine);
            }
            return sb.ToString();
        }
    }
}

File: semestr-8/Programowanie obiektowe w Java/alexander/Java/2020/ID06TC2/MatrixCalc/src/Matrix.java
import java.security.InvalidParameterException;

public class Matrix {
    private double[][] tb;
    private Matrix(double[][] tb) {
        this.tb = tb;
    }
    public Matrix(int rows, int cols) {
        tb = new double[rows][cols];
    }
    public int rows() {
        return tb.length;
    }
    public int cols() {
        return tb.length > 0 ? tb[0].length : 0;
    }
    public static Matrix parseMatrix(String source) {
        String[] Rows = source.split("[\r\n]+");
        int rows = Rows.length;
        int cols = 0;
        String[][] cells = new String[rows][];
        for (int r = 0; r < rows; ++r) {
            cells[r] = Rows[r].split("[\t\\s]+");
            if (r == 0) {
                cols = cells[r].length;
            } else if (cols != cells[r].length) {
                throw new InvalidParameterException("Nieregularna tablica.");
            }
        }
        double[][] m = new double[rows][cols];
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                m[r][c] = Double.parseDouble(cells[r][c].replace(',','.'));
            }
        }
        return new Matrix(m);
    }
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        int rows = this.rows();
        int cols = this.cols();
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (c > 0) {
                    sb.append('\t');
                }
                sb.append(String.format("%.2f", tb[r][c]));
            }
            sb.append(System.lineSeparator());
        }
        return sb.toString();
    }
    public static Matrix multiply(Matrix a, Matrix b) {
        int height = b.rows();
        if (height != a.cols()) throw new InvalidParameterException("Niepasujace rozmiary.");
        
        int rows = a.rows();
        int cols = b.cols();
        
        double[][] m= new double[rows][cols];
        
        for (int r=0;r<rows;++r)
        {
            for (int c=0;c<cols;++c)
            {
                double sum=0;
                for(int h = 0; h < height; ++h)
                {
                    sum+= a.tb[r][h]* b.tb[h][c];
                }
                m[r][c]=sum;
            }
        }
        return new Matrix(m);
    }
    public static Matrix gauss(Matrix a, Matrix b) {
        int rows = a.rows();
        if (rows != b.rows()) throw new InvalidParameterException("Niepasujace rozmiary.");
        int cols = a.cols();
        int add = b.cols();
        int size = cols+add;
        
        double[][] m = new double [rows][size];
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                m[r][c] = a.tb[r][c];
            }
            for (int c = 0; c < add; ++c) {
                m[r][c+cols] = b.tb[r][c];
            }
        }        
        for (int d = 0; d < rows; ++d) {
            double div = m[d][d];
            for (int r = 0; r < rows; ++r) {
                if(r != d) {
                    double mul = m[r][d];
                    for (int c = 0; c < size; ++c) {
                        m[r][c] -= m[d][c]*mul / div;
                    }
                }
            }
        }
        for (int r = 0; r < rows; ++r) {
            double div = m[r][r];
            for (int c = 0; c < add; ++c) {
                m[r][c] /= div;
            }
        }
        double[][] ret = new double[rows][add];
         for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < add; ++c) {
                ret[r][c] = m[r][c+cols];
            }
        }
        return new Matrix(ret);
    }
}
